### PR TITLE
MCO-358: inplaceupgrader: switch to using a daemonset for MCD

### PR DESCRIPTION
**Pull request by yuqi-zhang**
Today the inplace upgrader directly spins up pods for the
machine-config-daemons responsible for doing the inplace upgrade. This
was done for earlier versions of the logic where the pod was a oneshot,
but now the MCD is a state machine for Hypershift as well, so that is no
longer needed. Doing it as a pod has the following downsides:

1. the pods must be individually monitored for status
2. the pod must be referenced by the drainer so as not to drain it
   during upgrades
3. this deviates from self-driving OCP MCD behaviour

This PR aims to switch the MCD over to be a daemonset instead, with the
main downside that we now have to specify the SecurityContextConstraints
otherwise the pod cannot be created by the daemonset controller, meaning
that we will need a ServiceAccount, ClusterRole, and ClusterRoleBinding.

The other way I've found that can bypass the SecurityContextConstraints is to have a
"run-level" of 1 for the namespace, but that is not recommended. If there
are other options, I'm happy to change such that we don't need the
additional manifests.

cc @enxebre

### CODE REVIEW

**Comment by yuqi-zhang** [comment]
Rebased and fixed the selector so the daemonsets will run per nodepool. I've tested this to work with multiple upgrades and nodepools, so marking it ready for review.

The tradeoff here is that we need a bunch of additional objects to support the daemonset, with the benefits listed above.

**Comment by yuqi-zhang** [comment]
/retest

@enxebre would you have some time to take a look?

**Comment by yuqi-zhang** [comment]
```
=== CONT  TestInPlaceUpgradeNodePool
    util.go:328: Observed nodepool e2e-clusters-s4zmh/example-6hzmj-us-east-1c to report version 4.12.0-0.ci.test-2022-10-04-171253-ci-op-qs3zsjg6-latest in 2m0s
    util.go:709: Waiting for nodepool example-6hzmj-us-east-1c conditions to not be present: [UpdatingVersion]
--- FAIL: TestInPlaceUpgradeNodePool (1976.40s)
```

Hmm, did it never become ready?

/test e2e-aws

**Comment by yuqi-zhang** [comment]
Oh I guess the actual failure is:

```
fixture.go:119: Failed to dump cluster: failed to dump hosted cluster: found WARNING: Object got updated more than one time without a no-op update, this indicates hypershift incorrectly reverting defaulted values messages in file /logs/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/namespaces/e2e-clusters-r7vxh-example-9rk9d/core/pods/logs/hosted-cluster-config-operator-76985d985b-7fnp7-hosted-cluster-config-operator.log
```

which I can see in https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/1757/pull-ci-openshift-hypershift-main-e2e-aws/1577439036228767744/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/namespaces/e2e-clusters-r7vxh-example-9rk9d/core/pods/logs/hosted-cluster-config-operator-76985d985b-7fnp7-hosted-cluster-config-operator.log for the daemonset and the clusterrolebinding, hmm. Testing this locally didn't yield similar logs. I'm going to try to nuance the naming maybe?

File: control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go [hunk 1]
```diff
@@ -450,6 +388,105 @@ func (r *Reconciler) reconcileInPlaceUpgradeManifests(ctx context.Context, hoste
...
+	daemonset.Spec.Template.Spec.HostNetwork = true
+	daemonset.Spec.Template.Spec.HostPID = true
+	daemonset.Spec.Template.Spec.Tolerations = []corev1.Toleration{
+		{
+			Operator: corev1.TolerationOpExists,
+		},
+	}
+	daemonset.Spec.Template.Spec.ServiceAccountName = "machine-config-daemon"
+	daemonset.Spec.Template.Spec.NodeSelector = map[string]string{
+		"hypershift.openshift.io/nodePool": poolName,
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go:466]: can you please reference the constant declared for this annotation?

**Comment by yuqi-zhang** [comment]
I figured out the issue, after I switched all the manifests to per-namespace, there was one missing mapping that stalled the daemonset rollout.

Should hopefully pass the e2e test now. I tested the same flow locally and it worked

File: control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go [hunk 2]
```diff
@@ -456,6 +389,105 @@ func (r *Reconciler) reconcileInPlaceUpgradeManifests(ctx context.Context, hoste
...
+	daemonset.Spec.Template.Spec.HostNetwork = true
+	daemonset.Spec.Template.Spec.HostPID = true
+	daemonset.Spec.Template.Spec.Tolerations = []corev1.Toleration{
+		{
+			Operator: corev1.TolerationOpExists,
+		},
+	}
+	daemonset.Spec.Template.Spec.ServiceAccountName = "machine-config-daemon"
+	daemonset.Spec.Template.Spec.NodeSelector = map[string]string{
+		hyperv1.NodePoolLabel: poolName,
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go:467]: Does this mean the pod is running idle  in every Node for the NodePool until the Node is actually chosen to upgrade?

File: control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go [hunk 3]
```diff
@@ -609,3 +654,49 @@ func inPlaceUpgradeConfigMap(poolName string, namespace string) *corev1.ConfigMa
...
+}
+
+func inPlaceUpgradeDaemonClusterRole(namespace string) *rbacv1.ClusterRole {
+	return &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("machine-config-daemon-%s", namespace),
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups:     []string{"security.openshift.io"},
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go:674]: can you elaborate why is this needed now?

File: control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go (hunk 2 above)
Comment (enxebre) [reply control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go:467]: The daemonSet pod will actually get rescheduled even after the Node is upgraded until the whole NodePool is done and so the daemonSet deleted, right? I could think of envs with large NodePools that are "perpetually" upgrading where this means the pod is effectively long running. Could we somehow narrow scheduling to only the Nodes performing the upgrade within the NodePool?

File: control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go (hunk 3 above)
Comment (enxebre) [reply control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go:674]: I assume this is for a good reason which we should not bypass?

Comment (yuqi-zhang) [reply control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go:674]: This rule matches the self-driving OCP's MCD: https://github.com/openshift/machine-config-operator/blob/b6cfd42372427eb5d7aee44ce30c81c74ae12b08/manifests/machineconfigdaemon/clusterrole.yaml#L12-L15

The only way I am aware of bypassing it is to use `run-level: 1` which we removed from the MCO: https://github.com/openshift/machine-config-operator/commit/8c4dda1cc54bc85f951cd3a78bb7ba8a55a9f8bf

That seems even worse in terms of security, and I don't know of another way to do so

File: control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go (hunk 2 above)
Comment (yuqi-zhang) [reply control-plane-operator/hostedclusterconfigoperator/controllers/inplaceupgrader/inplaceupgrader.go:467]: Yes, that's definitely an optimisation we should try to make. I will experiment with more dynamic labelling to select only nodes that require the daemonset.

This isn't by itself worse than today, since the `pod` we dynamically schedule also doesn't yet have a cleanup process until the whole update completes. If its ok with you, I'd like to make that a separate card/PR, since I would want to do some extra testing to ensure we are not missing some cases where a node needs action but the controller doesn't schedule the pod

**Comment by yuqi-zhang** [comment]
Sure, I will spin that out to see what that looks like

**Comment by yuqi-zhang** [comment]
Sorry for the delay, I've updated this PR to only schedule the daemonset pod on nodes that require the update, via a label that the daemonset nodeselects for.

**Comment by yuqi-zhang** [comment]
@enxebre would you have some time to look at this recently?

**Comment by yuqi-zhang** [comment]
Let's revisit this at a later date. Let me know if there is any interest in pursuing using the daemonset flow. Thanks!
